### PR TITLE
Rewrite Helmsman to support Helm v3

### PR DESCRIPTION
**Pull request by mkubaczyk**
#274 
It's been humongous work so far, but we're close to the end I'd say.
Tiller-related parts were removed, logging has been reconfigured to follow some standard, changed lots of messages and aligned the logic of helmsman to helm 3 (since some things have changed).
Tests were also updated.
I've been testing it with rc1 of helm 3. Seems to work for both helm-secrets and hiera-eyaml secrets.

With such work there's high certainty some bugs happened, so we should start reviewing the changes. :-)

### CODE REVIEW

**Comment by mkubaczyk** [comment]
FYI
docs, readme etc. were not touched

**Comment by mkubaczyk** [comment]
For tests to pass, we need to change the docker image with the new version I've already pushed on branch

**Comment by virgofx** [comment]
:+1: any way a release/beta can be built with v3 support?

File: dockerfile/dockerfile [hunk 1]
```diff
@@ -24,16 +24,16 @@ ARG KUBE_VERSION
 ARG HELM_VERSION
 
 ENV KUBE_VERSION ${KUBE_VERSION:-v1.14.8}
-ENV HELM_VERSION ${HELM_VERSION:-v2.15.0}
-ENV HELM_DIFF_VERSION ${HELM_DIFF_VERSION:-v2.11.0+5}
+ENV HELM_VERSION ${HELM_VERSION:-v3.0.0-rc.1}
```
Comment (epierotto) [inline dockerfile/dockerfile]: Helm 3 is out, maybe the `rc` part can be removed from `HELM_VERSION`

File: dockerfile/dockerfile [hunk 2]
```diff
@@ -24,16 +24,16 @@ ARG KUBE_VERSION
...
+ENV HELM_DIFF_VERSION ${HELM_DIFF_VERSION:-v3.0.0-rc.7}
 
 RUN apk --no-cache update \
     && apk add --update --no-cache ca-certificates git openssh ruby \
     && apk add --update -t deps curl tar gzip make bash \
     && rm -rf /var/cache/apk/* \
     && curl -L https://storage.googleapis.com/kubernetes-release/release/${KUBE_VERSION}/bin/linux/amd64/kubectl -o /usr/local/bin/kubectl \
     && chmod +x /usr/local/bin/kubectl \
-    && curl -L http://storage.googleapis.com/kubernetes-helm/helm-${HELM_VERSION}-linux-amd64.tar.gz | tar zxv -C /tmp \
+    && curl -L https://get.helm.sh/helm-${HELM_VERSION}-linux-amd64.tar.gz | tar zxv -C /tmp \
```
Comment (epierotto) [inline dockerfile/dockerfile]: Just by tweaking a little bit the `curl` command maybe you can remove lines 37, 38 and 39
```
curl -L https://get.helm.sh/helm-${HELM_VERSION}-linux-amd64.tar.gz \
  | tar xzf - --strip-components=1 -C /usr/local/bin/ linux-amd64/helm
```
That will extract only the Helm binary into `/usr/local/bin/`

Actually the way it is now is more explicit than my command

File: aws/aws.go [hunk 3]
```diff
@@ -65,7 +65,7 @@ func ReadFile(bucketName string, filename string, outFile string, noColors bool)
 		log.Fatal(style.Bold(style.Red("ERROR: Failed to download file  " + filename + " from S3: " + err.Error())))
 	}
 
-	log.Println("INFO: Successfully downloaded " + filename + " from S3 as " + outFile)
+	log.Println("Successfully downloaded " + filename + " from S3 as " + outFile)
```
Comment (nvanheuverzwijn) [inline aws/aws.go]: Should this be `logs.Info` instead of `log.Println` ?

File: decision_maker.go [hunk 4]
```diff
@@ -225,31 +196,30 @@ func inspectUpgradeScenario(r *release, rs releaseState) {
...
+			logDecision("release [ "+r.Name+" ] is desired to use a new Chart [ "+r.Chart+
+				" ]. Delete of the current release will be planned and new chart will be installed in namespace [[ "+
+				r.Namespace+" ]]", r.Priority, change)
 		} else {
 			if diff := diffRelease(r); diff != "" {
 				upgradeRelease(r)
-				logDecision(generateDecisionMessage(r, "release [ "+r.Name+" ] is currently enabled and have some changed parameters. "+
-					"I will upgrade it!", false), r.Priority, change)
+				logDecision("release [ "+r.Name+" ] is installed and has some changes to apply. "+
+					"I will upgrade it!", r.Priority, change)
```
Comment (nvanheuverzwijn) [inline decision_maker.go]: I would remove `I will upgrade it!` from the log to be consistent with the refactor.

File: helm_helpers.go [hunk 5]
```diff
@@ -610,43 +323,36 @@ func cleanUntrackedReleases() {
 	}
 
 	if len(toDelete) == 0 {
-		log.Println("INFO: no untracked releases found.")
+		logs.Info("No untracked releases found.")
```
Comment (nvanheuverzwijn) [inline helm_helpers.go]: Upper-case first letter for logs seems inconsistent with other logs.Info

File: init.go [hunk 6]
```diff
@@ -113,7 +114,7 @@ func init() {
 	}
 
 	if len(files) == 0 {
-		log.Println("INFO: No desired state files provided.")
+		logs.Info("No desired state files provided.")
```
Comment (nvanheuverzwijn) [inline init.go]: Uppercase N seem inconsistent with other logs.Info

File: kube_helpers.go [hunk 7]
```diff
@@ -110,7 +56,7 @@ func labelNamespace(ns string, labels map[string]string) {
 
 		exitCode, _, _ := cmd.exec(debug, verbose)
 		if exitCode != 0 && verbose {
-			log.Println("WARN: I could not label namespace [ " + ns + " with " + k + "=" + v +
+			logs.Warning("I could not label namespace [ " + ns + " with " + k + "=" + v +
```
Comment (nvanheuverzwijn) [inline kube_helpers.go]: I would remove the "personnification" in the logs here to be consistent with log message refactoring. Maybe `namespace cannot be labeled [...]. It already exists. Skipping..`

File: kube_helpers.go [hunk 8]
```diff
@@ -127,7 +73,7 @@ func annotateNamespace(ns string, labels map[string]string) {
 
 		exitCode, _, _ := cmd.exec(debug, verbose)
 		if exitCode != 0 && verbose {
-			log.Println("WARN: I could not annotate namespace [ " + ns + " with " + k + "=" + v +
+			logs.Info("I could not annotate namespace [ " + ns + " with " + k + "=" + v +
```
Comment (nvanheuverzwijn) [inline kube_helpers.go]: I would remove the "personnification" in the logs here to be consistent with log message refactoring. Maybe `namespace cannot be annotated [...]. It already exists. Skipping..`

File: main.go [hunk 9]
```diff
@@ -109,12 +87,12 @@ func main() {
 			logError(msg)
 		}
 	} else {
-		log.Println("INFO: charts validation is skipped.")
+		logs.Info("Skipping charts' validation.")
```
Comment (nvanheuverzwijn) [inline main.go]: Uppercase S seems inconsistent with logs.Info

File: main.go [hunk 10]
```diff
@@ -109,12 +87,12 @@ func main() {
...
 	}
 
-	log.Println("INFO: checking what I need to do for your charts ... ")
+	logs.Info("Preparing plan...")
```
Comment (nvanheuverzwijn) [inline main.go]: Uppercase P seems inconsistent with logs.Info

File: main.go [hunk 11]
```diff
@@ -129,15 +107,13 @@ func main() {
...
-
-	log.Println("INFO: completed applying plan successfully!")
 }
 
 // cleanup deletes the k8s certificates and keys files
 // It also deletes any Tiller TLS certs and keys
 // and secret files
 func cleanup() {
-	log.Println("INFO: cleaning up sensitive and temp files")
+	logs.Info("Cleaning up sensitive and temp files.")
```
Comment (nvanheuverzwijn) [inline main.go]: Uppercase C seems inconsistent with logs.Info

File: plan.go [hunk 12]
```diff
@@ -88,30 +77,34 @@ func (p *plan) addDecision(decision string, priority int, decisionType decisionT
 func (p plan) execPlan() {
 	p.sortPlan()
 	if len(p.Commands) > 0 {
-		log.Println("INFO: Executing the plan ... ")
+		logs.Info("Executing the plan ... ")
```
Comment (nvanheuverzwijn) [inline plan.go]: Uppercase E seems inconsistent with logs.Info

File: plan.go [hunk 13]
```diff
@@ -88,30 +77,34 @@ func (p *plan) addDecision(decision string, priority int, decisionType decisionT
...
 	} else {
-		log.Println("INFO: Nothing to execute ... ")
+		logs.Info("Nothing to execute.")
```
Comment (nvanheuverzwijn) [inline plan.go]: Uppercase N seems inconsistent with logs.Info

File: state.go [hunk 14]
```diff
@@ -92,78 +90,39 @@ func (s state) validate() (bool, string) {
...
-			return false, "ERROR: certificates validation failed -- kube context setup is required but no certificates stanza provided."
+			return false, "certificates validation failed -- kube context setup is required but no certificates stanza provided."
 		}
 	}
 
 	// namespaces
 	if nsOverride == "" {
 		if s.Namespaces == nil || len(s.Namespaces) == 0 {
-			return false, "ERROR: namespaces validation failed -- I need at least one namespace " +
+			return false, "namespaces validation failed -- I need at least one namespace " +
```
Comment (nvanheuverzwijn) [inline state.go]: Would remove `I need at least one namespace to work with!` for `at least one namespace is required.`

File: state.go [hunk 15]
```diff
@@ -92,78 +90,39 @@ func (s state) validate() (bool, string) {
 
 		if s.Settings.ClusterURI != "" && !s.Settings.BearerToken {
 			if !caCrt || !caKey {
-				return false, "ERROR: certificates validation failed -- You want me to connect to your cluster for you " +
+				return false, "certificates validation failed -- You want me to connect to your cluster for you " +
```
Comment (nvanheuverzwijn) [inline state.go]: woudl remove the "personnification" here too.

File: kube_helpers.go [hunk 16]
```diff
@@ -413,17 +359,17 @@ func createRole(namespace string, role string, roleTemplateFile string) (bool, s
...
-			storageBackend = "secret"
+		if s.Settings.StorageBackend != "" {
+			storageBackend = s.Settings.StorageBackend
 		}
 
 		cmd := command{
 			Cmd:         "kubectl",
-			Args:        []string{"label", storageBackend, "-n", getDesiredTillerNamespace(r), "-l", "NAME=" + r.Name, "MANAGED-BY=HELMSMAN", "NAMESPACE=" + r.Namespace, "TILLER_NAMESPACE=" + getDesiredTillerNamespace(r), "--overwrite"},
-			Description: "applying labels to Helm state in [ " + getDesiredTillerNamespace(r) + " ] for " + r.Name,
+			Args:        []string{"label", storageBackend, "-l", "NAME=" + r.Name, "MANAGED-BY=HELMSMAN", "NAMESPACE=" + r.Namespace, "--overwrite"},
```
Comment (cihangirbesiktas) [inline kube_helpers.go]: You should run this command in the namespace of the release. So it's better to add `"-n", r.Namespace,` after `storageBackend,`.

**Comment by luisdavim** [comment]
@mkubaczyk I believe that all both the latest helm v2 and v3 versions support `-o json` in all commands that have an output, we should stop parsing the table outputs and use only the json outputs, that is much more robust and cleaner.

**Comment by mkubaczyk** [comment]
Thank you All for engagement into this. I just got back from my two weeks vacations and I’ll do my best to find time for this PR to be finished. :-) you can also propose changes directly with your code to this branch, this will help a lot

File: dockerfile/dockerfile (hunk 1 above)
Comment (mkubaczyk) [reply dockerfile/dockerfile]: Done 👍

File: dockerfile/dockerfile (hunk 2 above)
Comment (mkubaczyk) [reply dockerfile/dockerfile]: I'll let myself leave it as it is, don't want to spend time on sugar at this time, I'd rather like to reach RC state asap :-)

File: helm_helpers.go [hunk 17]
```diff
@@ -25,181 +24,78 @@ type releaseState struct {
...
-		// skipping the header from helm output
-		time, err := time.Parse("Mon Jan _2 15:04:05 2006", rel.Releases[i].Updated)
+	rel := getHelmReleases()
+
+	for i := 0; i < len(rel); i++ {
+		// we need to split the time into parts and make sure milliseconds len = 6, it happens to skip trailing zeros
+		updatedFields := strings.Fields(rel[i].Updated)
+		updatedHour := strings.Split(updatedFields[1], ".")
+		milliseconds := updatedHour[1]
+		for i := len(milliseconds); i < 6; i++ {
```
[the remark this answers is not in the document]
Comment (mkubaczyk) [reply helm_helpers.go]: I'll go for 9 then, thanks

File: decision_maker.go (hunk 4 above)
Comment (mkubaczyk) [reply decision_maker.go]: :+1:

File: helm_helpers.go (hunk 5 above)
Comment (mkubaczyk) [reply helm_helpers.go]: :+1:

File: init.go (hunk 6 above)
Comment (mkubaczyk) [reply init.go]: :+1:

File: kube_helpers.go (hunk 7 above)
Comment (mkubaczyk) [reply kube_helpers.go]: :+1:

File: kube_helpers.go (hunk 8 above)
Comment (mkubaczyk) [reply kube_helpers.go]: :+1:

File: main.go (hunk 9 above)
Comment (mkubaczyk) [reply main.go]: :+1:

File: main.go (hunk 11 above)
Comment (mkubaczyk) [reply main.go]: :+1:

File: plan.go (hunk 13 above)
Comment (mkubaczyk) [reply plan.go]: :+1:

File: state.go (hunk 14 above)
Comment (mkubaczyk) [reply state.go]: :+1:

File: state.go (hunk 15 above)
Comment (mkubaczyk) [reply state.go]: :+1:

File: kube_helpers.go (hunk 16 above)
Comment (mkubaczyk) [reply kube_helpers.go]: thanks :-)

**Comment by mkubaczyk** [comment]
> @mkubaczyk I believe that all both the latest helm v2 and v3 versions support `-o json` in all commands that have an output, we should stop parsing the table outputs and use only the json outputs, that is much more robust and cleaner.

Hi, will you please point specific places in code? I've changed two, but I might have missed something. :-)

File: state.go [hunk 18]
```diff
@@ -44,35 +42,35 @@ func (s state) validate() (bool, string) {
 
 	// settings
 	if (s.Settings == (config{}) || s.Settings.KubeContext == "") && !getKubeContext() {
-		return false, "ERROR: settings validation failed -- you have not defined a " +
+		return false, "settings validation failed -- you have not defined a " +
```
Comment (luisdavim) [inline state.go]: maybe we should just make this return an `error` instead of `(bool, string)`

Comment (mkubaczyk) [reply state.go]: that is a valid point, I've push commit with changes :-)

File: aws/aws.go (hunk 3 above)
Comment (mkubaczyk) [reply aws/aws.go]: since this is inside of aws package, I can't import logs from root in a clean way, so decided to let them be for now as they are

**Comment by mkubaczyk** [comment]
@luisdavim @sami-alajrami and others, with latest commit I've changed the structure of the repository in order to have *.go files separated from configuration files and so on.
A completely new logger was introduced.
I've also changed the circleci so we can have tests run based on PR-specific Dockerfile with no need to push image beforehand. Solution isn't ideal, but it unblocks work with tests where there are breaking changes introduced.
I'm more than convinced we can find bugs in my code, but I'd like to ask for one final review (might be just a glance) of this PR and then merge it to master. Current master was pushed as 1.x branch. Once we're on master, we can start fixing things with small PRs so it's easier to review and thanks to this we can have people testing our beta version of Helmsman 3 as soon as possible. :-)

**Comment by cihangirbesiktas** [comment]
It's failed when I tried `make cross`, 
`$ make cross
Error reading packages: exit status 1
Stderr: can't load package: package github.com/Praqma/helmsman: no Go files in /Users/cbesiktas/go/src/github.com/Praqma/helmsman
make: *** [cross] Error 1`

**Comment by cihangirbesiktas** [comment]
I got this error when I tried to upgrade an app 
```2019-12-17 16:50:12 CRITICAL: Command returned [ 1 ] exit code and error message [ Error: cannot re-use a name that is still in use```

and I am unable to delete an app when it's disabled. 
```
# ../helmsman-helm-v3 -f common.yaml -e common.env -f qa/apps.yaml -e qa/apps.env --no-fancy --apply
2019-12-17 16:49:15 INFO: Parsed YAML [[ common.yaml ]] successfully and found [ 0 ] apps
2019-12-17 16:49:15 INFO: Parsed YAML [[ qa/apps.yaml ]] successfully and found [ 1 ] apps
2019-12-17 16:49:26 INFO: Preparing plan...
2019-12-17 16:49:26 INFO: Acquiring current Helm state from cluster...
2019-12-17 16:49:27 INFO: Checking if any Helmsman managed releases are no longer tracked by your desired state ...
2019-12-17 16:49:28 INFO: No untracked releases found
2019-12-17 16:49:28 NOTICE: -------- PLAN starts here --------------
2019-12-17 16:49:28 INFO: Release [ twx-qa ] disabled -- priority: 0
2019-12-17 16:49:28 NOTICE: -------- PLAN ends here --------------
2019-12-17 16:49:28 INFO: Nothing to execute
```

**Comment by mkubaczyk** [comment]
@cihangirbesiktas 
regarding `make cross` I know what's wrong, will fix that.

regarding failing update - what says diff? Does this really want to change the app name?

regarding the deletion - so you have app deployed, then you're setting it to `enabled: false` and it doesn't want to delete it?

**Comment by cihangirbesiktas** [comment]
@mkubaczyk 
Regarding failing update, this is helmsman output 
```
2019-12-18 11:37:35 INFO: Preparing plan...
2019-12-18 11:37:35 INFO: Acquiring current Helm state from cluster...
2019-12-18 11:37:35 INFO: Checking if any Helmsman managed releases are no longer tracked by your desired state ...
2019-12-18 11:37:36 INFO: No untracked releases found
2019-12-18 11:37:36 NOTICE: -------- PLAN starts here --------------
2019-12-18 11:37:36 NOTICE: Release [ thingworx ] will be installed in [ dev ] namespace -- priority: 0
2019-12-18 11:37:36 NOTICE: -------- PLAN ends here --------------
2019-12-18 11:37:36 INFO: Executing plan...
2019-12-18 11:37:36 NOTICE: Installing release [ thingworx ] in namespace [ dev ]
2019-12-18 11:37:37 CRITICAL: Command returned [ 1 ] exit code and error message [ Error: cannot re-use a name that is still in use
```

regarding the deletion - yes, that's what happened.

**Comment by mkubaczyk** [comment]
@cihangirbesiktas 
so when it comes to the deletion - I can't reproduce it. I was deploying apps, then removing it and so on and everything went fine. Maybe we're missing some important point here

the update - any chance that helm 3 had such chart deployed before you run helmsman so it could not deploy second one with the same name? might be that acquiring state is considering such chart.
that can happen where there's thingworx chart installed but it was not labeled by Helmsman, so when it tries to get current state of releases, it misses him

File: internal/app/cli.go [hunk 19]
```diff
@@ -159,49 +147,44 @@ func init() {
...
+			if err := s.validate(); err != nil { // syntax validation
+				log.Error(err.Error())
 			}
 		}
 	} else {
-		log.Println("INFO: desired state validation is skipped.")
+		log.Info("Desired state validation is skipped.")
 	}
 
 	if applyLabels {
```
Comment (sami-alajrami) [inline internal/app/cli.go]: the `apply labels` option should be deprecated. It was introduced to make it possible to migrate from one version to another in v1.x. This should no longer be needed here.

File: .circleci/config.yml [hunk 20]
```diff
@@ -44,27 +28,19 @@ jobs:
           command: |
             TAG=$(git describe --abbrev=0 --tags)
             docker login -u $DOCKER_USER -p $DOCKER_PASS
-            docker build -t praqma/helmsman:$TAG-helm-v2.15.1 --build-arg HELM_VERSION=v2.15.1 --build-arg HELM_DIFF_VERSION=master dockerfile/. --no-cache
```
Comment (sami-alajrami) [inline .circleci/config.yml]: the following line should be removed and replaced with ones for helm 3

File: internal/app/bindata.go [hunk 21]
```diff
@@ -3,7 +3,7 @@
 // data/role.yaml
 // DO NOT EDIT!
 
-package main
```
Comment (sami-alajrami) [inline internal/app/bindata.go]: This file is no longer needed together with `data/role.yaml` as it was only needed for Tiller.

File: internal/app/state.go [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+// state type represents the desired state of applications on a k8s cluster.
+type state struct {
+	Metadata               map[string]string    `yaml:"metadata"`
+	Certificates           map[string]string    `yaml:"certificates"`
+	Settings               config               `yaml:"settings"`
+	Namespaces             map[string]namespace `yaml:"namespaces"`
+	HelmRepos              map[string]string    `yaml:"helmRepos"`
+	PreconfiguredHelmRepos []string             `yaml:"preconfiguredHelmRepos"`
+	Apps                   map[string]*release  `yaml:"apps"`
+	AppsTemplates          map[string]*release  `yaml:"appsTemplates,omitempty"`
```
[the remark this answers is not in the document]
Comment (mkubaczyk) [reply internal/app/state.go]: that was it #302

File: internal/app/kube_helpers.go [hunk 23]
```diff
@@ -495,14 +440,12 @@ func getHelmsmanReleases() map[string]map[string]bool {
 			if r == "" {
 				continue
 			}
-			r = regexp.MustCompile(`(^\w+\/|\.v\d+$)`).ReplaceAllString(r, "")
 			if _, ok := releases[ns]; !ok {
-				releases[ns] = make(map[string]bool)
+				releases[ns] = make(map[*release]bool)
 			}
-			releases[ns][r] = false
 			for _, app := range s.Apps {
```
[the remark this answers is not in the document]
Comment (mkubaczyk) [reply internal/app/kube_helpers.go]: would be a small relief for me if you take it :-)

**Comment by mkubaczyk** [comment]
@sami-alajrami thank you for such comprehensive review. Applied all changes you asked and left one thing regarding detecting untracked releases for you (thank you for this too). If you're fine with newest changes, feel free to merge MR so we can start working on smaller, reviewable things for helmsman 3 :-)

**Comment by mkubaczyk** [comment]
@sami-alajrami done, just move the return outside the if statement

**Pull request merged**